Sources/MongoSwift/MongoClient.swift

```diff
@@ -322,11 +317,33 @@ public class MongoClient {
      */
     public func withSession<T>(
         options: ClientSessionOptions? = nil,
-        _ sessionBody: (ClientSession) throws -> T
-    ) throws -> T {
-        let session = try ClientSession(client: self, options: options)
-        defer { session.end() }
-        return try sessionBody(session)
+        _ sessionBody: (ClientSession) throws -> EventLoopFuture<T>
+    ) -> EventLoopFuture<T> {
+        let promise = self.operationExecutor.makePromise(of: T.self)
+        let session = self.startSession(options: options)
+        do {
+            let bodyFuture = try sessionBody(session)
+            // regardless of whether body's returned future succeeds we want to call session.end() once its complete.
+            // only once session.end() finishes can we fulfill the returned promise. otherwise the user can't tell if
+            // it is safe to close the parent client of this session, and they could inadvertently close it before the
+            // session is actually ended and its parent `mongoc_client_t` is returned to the pool.
+            bodyFuture
```

Comment:
I think we do need to wait for the session to finish ending before we give the user a result, per my comment above this line - we need a way for the user to know it's safe to close the client they are calling this method on.

I think what we have here where we automatically do the conversion for them by catching the error and failing the promise is probably easier so they don't really have to think about it. they might not need it that often and it would probably just be for code that could have been called outside of the closure anyway, like encoding something, but for convenience now they can do it if they want to.

Reply:
Sounds good.

One thing to note, if `session.end()` could fail (which it can't), then the `flatMapError` would retry the same thing that just failed. Just something to keep in mind if we continue this pattern further. I guess a solution could be to `recover` the first one, though we'd lose the error from `close`.

If we plan to repeat this pattern elsewhere, we could add an internal extension to `EventLoopFuture` to encapsulate this chain in a single callback (I did something simimlar in my cursor PR):
```swift
extension EventLoopFuture {
    internal func completeAfter(promise: EventLoopPromise<Value>, f: @escaping () -> EventLoopFuture<Void>) -> EventLoopFuture<Value> {
        self.flatMap { _ in 
            f().recover { _ in () }
        }.flatMapError { _ in
            return f()
        }.whenComplete { promise.completeWith(self) }
    }
}
```
Then we could use it here like:
```swift
try sessionBody(session).completeAfter(promise: promise) { session.close() }
```
`completeAfter` isn't the best name in the world, but I couldn't really come up with anything better at the moment. It's also possible that this functionality would be better as an extension of `EventLoopPromise` and named `completeWithAfter` or something, e.g.
```swift
promise.completeWithAfter(try sessionBody(session)) { session.close() }
```
Again, this extension part probably isn't necessary unless we plan to reproduce this pattern in the future.

Reply:
off the top of my head I can't really think of any places we need to use this, but will keep in mind in case any more arise in the rest of the implementation or tests 👍